File: apps/scraper/app/strategies.py
import json
import logging
import re

from bs4 import BeautifulSoup, Tag

from .parsing import (
    ExtractionResult,
    _extract_price_from_text,
    _extract_stock_from_text,
    _is_element_disabled,
    _normalize_text,
    _parse_price_match,
)

logger = logging.getLogger(__name__)


# --- CSS selector extraction ---


_INTERACTIVE_TAGS = {"button", "input", "a"}


def _extract_css_selector(soup: BeautifulSoup, css_selector: str) -> ExtractionResult | None:
    """Extract price/stock from a user-provided CSS selector."""
    elements = soup.select(css_selector)
    if not elements:
        return None
    text = " ".join(el.get_text(strip=True) for el in elements)
    price = _extract_price_from_text(text)
    stock_status = _extract_stock_from_text(text)
    if price is None and stock_status is None:
        return None
    # A disabled interactive element shouldn't imply in_stock
    if stock_status == "in_stock":
        for el in elements:
            is_interactive = (
                el.name in _INTERACTIVE_TAGS
                or el.get("role", "").lower() == "button"
            )
            if is_interactive and _is_element_disabled(el):
                stock_status = None
                break
    return ExtractionResult(price=price, stock_status=stock_status, raw=text)


# --- Title window extraction ---


_NOISE_KEYWORDS = re.compile(
    r"\b(shipping|delivery|handling|tax|estimated|free delivery)\b", re.IGNORECASE
)
_OLD_PRICE_PREFIXES = re.compile(
    r"^\s*(was|original|compare\s+at|reg\.?|msrp|list\s+price|regular\s+price)\s*:?\s*",
    re.IGNORECASE,
)


def _extract_title_window(soup: BeautifulSoup) -> ExtractionResult | None:
    """Inspect visible text around the product title."""
    heading = _find_primary_heading(soup)
    lines = [line.strip() for line in soup.stripped_strings if line and line.strip()]
    if not lines:
        return None

    if heading:
        normalized_heading = _normalize_text(heading)
        start_index = next(
            (
                index
                for index, line in enumerate(lines)
                if normalized_heading and (
                    normalized_heading in _normalize_text(line)
                    or _normalize_text(line) in normalized_heading
                )
            ),
            0,
        )
    else:
        start_index = 0

    window_lines = lines[start_index:start_index + 10]

    filtered = []
    for line in window_lines:
        if _NOISE_KEYWORDS.search(line):
            continue
        if _OLD_PRICE_PREFIXES.match(line):
            continue
        filtered.append(line)

    if not filtered:
        return None

    window_text = " | ".join(filtered)
    price = _extract_price_from_text(window_text)
    stock_status = _extract_stock_from_text(window_text)

    if price is None and stock_status is None:
        return None

    return ExtractionResult(price=price, stock_status=stock_status, raw=window_text[:300])


def _find_primary_heading(soup: BeautifulSoup) -> str | None:
    heading = soup.find("h1")
    if heading:
        text = heading.get_text(strip=True)
        if text:
            return text

    og_title = soup.find("meta", attrs={"property": "og:title"})
    if og_title and og_title.get("content"):
        return og_title["content"]

    if soup.title and soup.title.string:
        return soup.title.string.strip()

    return None


# --- Common CSS selectors ---


def _extract_common_selectors(soup: BeautifulSoup) -> ExtractionResult | None:
    """Try common CSS selectors for price and stock."""
    price, price_raw = _extract_price_selectors(soup)
    stock_status, stock_raw = _extract_stock_selectors(soup)
    if price is None and stock_status is None:
        return None
    raw_parts = [r for r in (price_raw, stock_raw) if r]
    return ExtractionResult(
        price=price,
        stock_status=stock_status,
        raw=", ".join(raw_parts) if raw_parts else None,
    )


def _extract_price_selectors(soup: BeautifulSoup) -> tuple[float | None, str | None]:
    """Try common CSS selectors to find prices."""
    selectors = [
        '[data-price]',
        '[itemprop="price"]',
        '#product .price',
        '.product-detail .price',
        'main .price',
        '[data-product-price]',
        '.product__price',
        '.price__current',
        '.price-current',
        '.product-price',
        '.price .money',
        '.price-box .price',
        '.product-info-price .price',
        '#priceblock_ourprice',
        '#priceblock_dealprice',
        '.a-price .a-offscreen',
        '.price',
    ]
    for selector in selectors:
        elements = soup.select(selector)
        for el in elements:
            data_price = el.get("data-price") or el.get("data-product-price")
            if data_price:
                try:
                    val = float(data_price)
                    if val > 0:
                        return val, f"{selector}: data-price={data_price}"
                except ValueError:
                    pass

            content = el.get("content")
            if content:
                try:
                    val = float(content)
                    if val > 0:
                        return val, f"{selector}: content={content}"
                except ValueError:
                    pass

            text = el.get_text(strip=True)
            if text:
                price = _extract_price_from_text(text)
                if price is not None:
                    return price, f"{selector}: {text}"
    return None, None


def _extract_stock_selectors(soup: BeautifulSoup) -> tuple[str | None, str | None]:
    """Try to determine stock status from common selectors."""
    selectors = [
        '[itemprop="availability"]',
        '.availability',
        '.stock-status',
        '#availability',
        '.product-availability',
    ]
    for selector in selectors:
        elements = soup.select(selector)
        for el in elements:
            content = (el.get("content") or el.get("href") or "").lower()
            text = el.get_text(strip=True).lower()
            combined = f"{content} {text}"

            # Check schema.org URL patterns directly
            if "instock" in content:
                return "in_stock", f"{selector}: {content}"
            if "outofstock" in content:
                return "out_of_stock", f"{selector}: {content}"

            stock_status = _extract_stock_from_text(combined)
            if stock_status is not None:
                return stock_status, f"{selector}: {text or content}"

    # Generic add-to-cart button detection (late-pipeline fallback)
    add_to_cart_selectors = [
        "button[name='add-to-cart']",
        "[data-testid='add-to-cart']",
        "#addToCart",
        "#add-to-cart",
        ".add-to-cart",
        "button.add-to-cart-button",
        "form[action*='/cart'] button[type='submit']",
    ]
    for selector in add_to_cart_selectors:
        el = soup.select_one(selector)
        if el is not None:
            if _is_element_disabled(el):
                return "out_of_stock", f"{selector}: disabled"
            return "in_stock", f"{selector}: enabled"

    # Text-based button detection as final fallback
    return _extract_stock_from_buttons(soup)


_ADD_TO_CART_TEXTS = {"add to cart", "add to bag", "add to basket", "buy now", "buy it now"}
_OOS_BUTTON_TEXTS = {"sold out", "notify me", "join waitlist", "coming soon"}


def _extract_stock_from_buttons(soup: BeautifulSoup) -> tuple[str | None, str | None]:
    """Detect stock status from button/link visible text (no CSS class needed)."""
    elements = soup.find_all(["button", "input", "a"], limit=50)

    atc_element = None
    oos_element = None

    for el in elements:
        if el.name == "input":
            text = (el.get("value") or "").strip()
        else:
            text = el.get_text(strip=True)

        if not text or len(text) > 80:
            continue

        normalized = text.lower()

        if any(pat in normalized for pat in _ADD_TO_CART_TEXTS):
            atc_element = el
            break  # strongest signal, stop scanning
        if oos_element is None and any(pat in normalized for pat in _OOS_BUTTON_TEXTS):
            oos_element = el

    if atc_element is not None:
        if _is_element_disabled(atc_element):
            return "out_of_stock", f"button text: disabled"
        return "in_stock", f"button text: enabled"

    if oos_element is not None:
        return "out_of_stock", f"button text: oos"

    return None, None


# --- Element fingerprint extraction ---


_STABLE_ATTRS = [
    "data-testid", "data-test-id", "data-qa", "data-cy",
    "data-product-id", "data-price", "data-sku",
    "itemprop", "role", "type", "name", "aria-label",
]

_PRICE_PATTERN = re.compile(r"[\$\€\£¥₹]\s*\d|^\d[\d,]*\.\d{2}$")


def _normalize_ws(text: str) -> str:
    return re.sub(r"\s+", " ", text).strip().lower()


def _text_overlaps(stored: str, candidate: str) -> bool:
    """Check if texts overlap meaningfully. For price elements, match on price pattern presence."""
    s = _normalize_ws(stored)
    c = _normalize_ws(candidate)
    if not s or not c:
        return False
    # If stored text looks like a price, just check that candidate also has a price-like pattern
    if _PRICE_PATTERN.search(s):
        return _PRICE_PATTERN.search(c) is not None
    # Otherwise use substring containment
    return s in c or c in s


def _ancestor_tags_score(stored: list[str], candidate_el: Tag) -> int:
    """Score how well the candidate's ancestor chain matches the stored one."""
    ancestors: list[str] = []
    parent = candidate_el.parent
    while parent and parent.name not in ("[document]", "html") and len(ancestors) < 10:
        ancestors.insert(0, parent.name)
        parent = parent.parent
    # Count matching positions from the end (closest ancestors matter most)
    score = 0
    for s, c in zip(reversed(stored), reversed(ancestors)):
        if s == c:
            score += 1
        else:
            break
    return score


def _result_from_element(el: Tag) -> ExtractionResult | None:
    text = el.get_text(strip=True)
    if not text:
        return None
    price = _extract_price_from_text(text)
    stock_status = _extract_stock_from_text(text)
    if price is None and stock_status is None:
        return None
    return ExtractionResult(price=price, stock_status=stock_status, raw=text)


def _extract_fingerprint(soup: BeautifulSoup, fingerprint_json: str) -> ExtractionResult | None:
    """Extract price/stock using an element fingerprint as fallback."""
    try:
        fp = json.loads(fingerprint_json)
    except (json.JSONDecodeError, TypeError):
        return None

    tag_name = fp.get("tagName", "").lower()
    stored_text = fp.get("textContent", "")
    attributes = fp.get("attributes", {})
    ancestor_tags = fp.get("ancestorTags", [])
    nearest_id = fp.get("nearestIdAncestor")
    nearest_heading = fp.get("nearestHeading")

    if not tag_name:
        return None

    # 1. Stable attribute match -- tagName + any stored stable attribute
    for attr in _STABLE_ATTRS:
        val = attributes.get(attr)
        if val:
            candidates = soup.find_all(tag_name, attrs={attr: val})
            if len(candidates) == 1:
                result = _result_from_element(candidates[0])
                if result:
                    logger.info("  [fingerprint] matched via %s=%s", attr, val)
                    return result
            elif candidates:
                # Multiple matches -- try to narrow by text
                for c in candidates:
                    if _text_overlaps(stored_text, c.get_text(strip=True)):
                        result = _result_from_element(c)
                        if result:
                            logger.info("  [fingerprint] matched via %s=%s + text", attr, val)
                            return result

    # 2. ID-anchored text match
    if nearest_id:
        anchor = soup.find(id=nearest_id)
        if anchor:
            descendants = anchor.find_all(tag_name)
            for el in descendants:
                if _text_overlaps(stored_text, el.get_text(strip=True)):
                    result = _result_from_element(el)
                    if result:
                        logger.info("  [fingerprint] matched via id-anchor #%s + text", nearest_id)
                        return result

    # 3. Heading-anchored text match
    if nearest_heading:
        normalized_heading = _normalize_ws(nearest_heading)
        for h_tag in ["h1", "h2", "h3", "h4", "h5", "h6"]:
            for h in soup.find_all(h_tag):
                if normalized_heading in _normalize_ws(h.get_text(strip=True)):
                    # Search siblings and nearby elements
                    parent = h.parent
                    if parent:
                        for el in parent.find_all(tag_name):
                            if _text_overlaps(stored_text, el.get_text(strip=True)):
                                result = _result_from_element(el)
                                if result:
                                    logger.info("  [fingerprint] matched via heading-anchor + text")
                                    return result

    # 4. Global text match -- find all tagName elements with matching text, prefer closest ancestor match
    all_candidates = soup.find_all(tag_name)
    text_matches: list[tuple[int, Tag]] = []
    for el in all_candidates:
        if _text_overlaps(stored_text, el.get_text(strip=True)):
            score = _ancestor_tags_score(ancestor_tags, el) if ancestor_tags else 0
            text_matches.append((score, el))

    if text_matches:
        text_matches.sort(key=lambda x: x[0], reverse=True)
        result = _result_from_element(text_matches[0][1])
        if result:
            logger.info("  [fingerprint] matched via global text (ancestor score=%d)", text_matches[0][0])
            return result

    return None


# --- Meta tag extraction ---


_META_PRICE_PROPERTIES = [
    "product:price:amount",
    "og:price:amount",
    "product:price",
    "og:price",
    "price",
]


def _extract_meta_price(soup: BeautifulSoup) -> ExtractionResult | None:
    """Extract price from meta tags, checked in priority order."""
    for target_prop in _META_PRICE_PROPERTIES:
        for meta in soup.find_all("meta"):
            prop = (meta.get("property", "") or meta.get("name", "")).lower()
            if prop == target_prop:
                content = meta.get("content", "")
                if content:
                    cleaned = re.sub(r"[^\d.,]", "", content)
                    price = _parse_price_match(cleaned)
                    if price is not None and price > 0:
                        return ExtractionResult(price=price, stock_status=None, raw=f"meta[{prop}]={content}")
    return None

File: apps/scraper-rs/src/strategies.rs
use once_cell::sync::Lazy;
use regex::Regex;
use scraper::{ElementRef, Html};
use serde_json::Value;

use crate::{
    dom::{all_elements_by_tag, select_all, select_first, select_from_element},
    parsing::{
        ExtractionResult, extract_price_from_text, extract_stock_from_text, is_element_disabled,
        is_purchase_cta_text, normalize_text, parse_price_match, text_of_element,
    },
};

const INTERACTIVE_TAGS: &[&str] = &["button", "input", "a"];
const STABLE_ATTRS: &[&str] = &[
    "data-testid",
    "data-test-id",
    "data-qa",
    "data-cy",
    "data-product-id",
    "data-price",
    "data-sku",
    "itemprop",
    "role",
    "type",
    "name",
    "aria-label",
];
const ADD_TO_CART_TEXTS: &[&str] = &[
    "add to cart",
    "add to bag",
    "add to basket",
    "buy now",
    "buy it now",
];
const OOS_BUTTON_TEXTS: &[&str] = &["sold out", "notify me", "join waitlist", "coming soon"];
const META_PRICE_PROPERTIES: &[&str] = &[
    "product:price:amount",
    "og:price:amount",
    "product:price",
    "og:price",
    "price",
];

static NOISE_KEYWORDS: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"\b(shipping|delivery|handling|tax|estimated|free delivery)\b")
        .expect("valid noise keyword regex")
});
static OLD_PRICE_PREFIXES: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^\s*(was|original|compare\s+at|reg\.?|msrp|list\s+price|regular\s+price)\s*:?\s*")
        .expect("valid old price regex")
});
static PRICE_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"[$€£¥₹]\s*\d|^\d[\d,]*\.\d{2}$").expect("valid price pattern"));

pub fn extract_css_selector(document: &Html, css_selector: &str) -> Option<ExtractionResult> {
    let elements = select_all(document, css_selector);
    if elements.is_empty() {
        return None;
    }

    let text = elements
        .iter()
        .map(text_of_element)
        .collect::<Vec<_>>()
        .join(" ");
    let price = extract_price_from_text(&text);
    let mut stock_status = extract_stock_from_text(&text);

    if price.is_none() && stock_status.is_none() {
        return None;
    }

    if stock_status.as_deref() == Some("in_stock") {
        for element in &elements {
            let tag_name = element.value().name();
            let is_interactive = INTERACTIVE_TAGS.contains(&tag_name)
                || element
                    .value()
                    .attr("role")
                    .is_some_and(|value| value.eq_ignore_ascii_case("button"));
            let element_text = if tag_name == "input" {
                element
                    .value()
                    .attr("value")
                    .unwrap_or_default()
                    .to_string()
            } else {
                text_of_element(element)
            };
            if is_interactive && is_element_disabled(element) && is_purchase_cta_text(&element_text)
            {
                stock_status = Some("out_of_stock".to_string());
                break;
            }
        }
    }

    Some(ExtractionResult {
        price,
        stock_status,
        raw: Some(text),
    })
}

pub fn extract_title_window(document: &Html) -> Option<ExtractionResult> {
    let heading = find_primary_heading(document);
    let lines = document
        .root_element()
        .text()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .map(str::to_string)
        .collect::<Vec<_>>();
    if lines.is_empty() {
        return None;
    }

    let start_index = heading
        .as_deref()
        .map(normalize_text)
        .and_then(|normalized_heading| {
            lines.iter().position(|line| {
                let normalized_line = normalize_text(line);
                (!normalized_heading.is_empty())
                    && (normalized_line.contains(&normalized_heading)
                        || normalized_heading.contains(&normalized_line))
            })
        })
        .unwrap_or(0);

    let filtered = lines[start_index..usize::min(start_index + 10, lines.len())]
        .iter()
        .filter(|line| !NOISE_KEYWORDS.is_match(&line.to_lowercase()))
        .filter(|line| !OLD_PRICE_PREFIXES.is_match(line))
        .cloned()
        .collect::<Vec<_>>();

    if filtered.is_empty() {
        return None;
    }

    let window_text = filtered.join(" | ");
    let price = extract_price_from_text(&window_text);
    let stock_status = extract_stock_from_text(&window_text);

    if price.is_none() && stock_status.is_none() {
        return None;
    }

    Some(ExtractionResult {
        price,
        stock_status,
        raw: Some(window_text.chars().take(300).collect()),
    })
}

pub fn extract_common_selectors(document: &Html) -> Option<ExtractionResult> {
    let (price, price_raw) = extract_price_selectors(document);
    let (stock_status, stock_raw) = extract_stock_selectors(document);

    if price.is_none() && stock_status.is_none() {
        return None;
    }

    let raw_parts = [price_raw, stock_raw]
        .into_iter()
        .flatten()
        .collect::<Vec<_>>();

    Some(ExtractionResult {
        price,
        stock_status,
        raw: (!raw_parts.is_empty()).then(|| raw_parts.join(", ")),
    })
}

pub fn extract_price_selectors(document: &Html) -> (Option<f64>, Option<String>) {
    let selectors = [
        "[data-price]",
        r#"[itemprop="price"]"#,
        "#product .price",
        ".product-detail .price",
        "main .price",
        "[data-product-price]",
        ".product__price",
        ".price__current",
        ".price-current",
        ".product-price",
        ".price .money",
        ".price-box .price",
        ".product-info-price .price",
        "#priceblock_ourprice",
        "#priceblock_dealprice",
        ".a-price .a-offscreen",
        ".price",
    ];

    for selector in selectors {
        for element in select_all(document, selector) {
            if let Some(data_price) = element
                .value()
                .attr("data-price")
                .or_else(|| element.value().attr("data-product-price"))
            {
                if let Ok(value) = data_price.parse::<f64>() {
                    if value > 0.0 {
                        return (
                            Some(value),
                            Some(format!("{selector}: data-price={data_price}")),
                        );
                    }
                }
            }

            if let Some(content) = element.value().attr("content") {
                if let Ok(value) = content.parse::<f64>() {
                    if value > 0.0 {
                        return (Some(value), Some(format!("{selector}: content={content}")));
                    }
                }
            }

            let text = text_of_element(&element);
            if let Some(price) = extract_price_from_text(&text) {
                return (Some(price), Some(format!("{selector}: {text}")));
            }
        }
    }

    (None, None)
}

pub fn extract_stock_selectors(document: &Html) -> (Option<String>, Option<String>) {
    let selectors = [
        r#"[itemprop="availability"]"#,
        ".availability",
        ".stock-status",
        "#availability",
        ".product-availability",
    ];

    for selector in selectors {
        for element in select_all(document, selector) {
            let content = element
                .value()
                .attr("content")
                .or_else(|| element.value().attr("href"))
                .unwrap_or_default()
                .to_lowercase();
            let text = text_of_element(&element).to_lowercase();
            let combined = format!("{content} {text}");

            if content.contains("instock") {
                return (
                    Some("in_stock".to_string()),
                    Some(format!("{selector}: {content}")),
                );
            }
            if content.contains("outofstock") {
                return (
                    Some("out_of_stock".to_string()),
                    Some(format!("{selector}: {content}")),
                );
            }

            if let Some(stock_status) = extract_stock_from_text(&combined) {
                return (
                    Some(stock_status),
                    Some(format!(
                        "{selector}: {}",
                        if text.is_empty() { content } else { text }
                    )),
                );
            }
        }
    }

    let add_to_cart_selectors = [
        "button[name='add-to-cart']",
        "[data-testid='add-to-cart']",
        "#addToCart",
        "#add-to-cart",
        ".add-to-cart",
        "button.add-to-cart-button",
        "form[action*='/cart'] button[type='submit']",
    ];

    for selector in add_to_cart_selectors {
        if let Some(element) = select_first(document, selector) {
            if is_element_disabled(&element) {
                return (
                    Some("out_of_stock".to_string()),
                    Some(format!("{selector}: disabled")),
                );
            }
            return (
                Some("in_stock".to_string()),
                Some(format!("{selector}: enabled")),
            );
        }
    }

    extract_stock_from_buttons(document)
}

pub fn extract_fingerprint(document: &Html, fingerprint_json: &str) -> Option<ExtractionResult> {
    let Ok(value) = serde_json::from_str::<Value>(fingerprint_json) else {
        return None;
    };

    let tag_name = value
        .get("tagName")
        .and_then(Value::as_str)
        .map(str::to_lowercase)
        .unwrap_or_default();
    if tag_name.is_empty() {
        return None;
    }

    let stored_text = value
        .get("textContent")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let attributes = value.get("attributes").and_then(Value::as_object);
    let ancestor_tags = value
        .get("ancestorTags")
        .and_then(Value::as_array)
        .map(|tags| {
            tags.iter()
                .filter_map(Value::as_str)
                .map(str::to_lowercase)
                .collect::<Vec<_>>()
        })
        .unwrap_or_default();
    let nearest_id = value.get("nearestIdAncestor").and_then(Value::as_str);
    let nearest_heading = value.get("nearestHeading").and_then(Value::as_str);

    if let Some(attributes) = attributes {
        for attr in STABLE_ATTRS {
            let Some(expected) = attributes.get(*attr).and_then(Value::as_str) else {
                continue;
            };

            let candidates = all_elements_by_tag(document, &tag_name)
                .into_iter()
                .filter(|element| {
                    element
                        .value()
                        .attr(attr)
                        .is_some_and(|value| value == expected)
                })
                .collect::<Vec<_>>();

            if candidates.len() == 1 {
                if let Some(result) = result_from_element(&candidates[0]) {
                    return Some(result);
                }
            } else {
                for candidate in candidates {
                    if text_overlaps(stored_text, &text_of_element(&candidate)) {
                        if let Some(result) = result_from_element(&candidate) {
                            return Some(result);
                        }
                    }
                }
            }
        }
    }

    if let Some(anchor_id) = nearest_id {
        if let Some(anchor) = select_first(document, &format!("#{anchor_id}")) {
            for element in select_from_element(&anchor, &tag_name) {
                if text_overlaps(stored_text, &text_of_element(&element)) {
                    if let Some(result) = result_from_element(&element) {
                        return Some(result);
                    }
                }
            }
        }
    }

    if let Some(heading) = nearest_heading {
        let normalized_heading = normalize_text(heading);
        for selector in ["h1", "h2", "h3", "h4", "h5", "h6"] {
            for candidate_heading in select_all(document, selector) {
                if normalize_text(&text_of_element(&candidate_heading))
                    .contains(&normalized_heading)
                {
                    for element in select_from_element(&candidate_heading, &tag_name) {
                        if text_overlaps(stored_text, &text_of_element(&element)) {
                            if let Some(result) = result_from_element(&element) {
                                return Some(result);
                            }
                        }
                    }

                    let mut parent = candidate_heading.parent();
                    while let Some(node) = parent {
                        if let Some(parent_element) = ElementRef::wrap(node) {
                            for element in select_from_element(&parent_element, &tag_name) {
                                if text_overlaps(stored_text, &text_of_element(&element)) {
                                    if let Some(result) = result_from_element(&element) {
                                        return Some(result);
                                    }
                                }
                            }
                            break;
                        }
                        parent = node.parent();
                    }
                }
            }
        }
    }

    let mut matches = all_elements_by_tag(document, &tag_name)
        .into_iter()
        .filter_map(|element| {
            text_overlaps(stored_text, &text_of_element(&element)).then(|| {
                let score = if ancestor_tags.is_empty() {
                    0
                } else {
                    ancestor_tags_score(&ancestor_tags, &element)
                };
                (score, element)
            })
        })
        .collect::<Vec<_>>();

    matches.sort_by(|left, right| right.0.cmp(&left.0));
    matches
        .into_iter()
        .find_map(|(_score, element)| result_from_element(&element))
}

pub fn extract_meta_price(document: &Html) -> Option<ExtractionResult> {
    for target_prop in META_PRICE_PROPERTIES {
        for meta in select_all(document, "meta") {
            let prop = meta
                .value()
                .attr("property")
                .or_else(|| meta.value().attr("name"))
                .unwrap_or_default()
                .to_lowercase();
            if prop != *target_prop {
                continue;
            }

            let Some(content) = meta.value().attr("content") else {
                continue;
            };
            if let Some(price) = parse_price_match(content) {
                return Some(ExtractionResult {
                    price: Some(price),
                    stock_status: None,
                    raw: Some(format!("meta[{prop}]={content}")),
                });
            }
        }
    }

    None
}

fn find_primary_heading(document: &Html) -> Option<String> {
    if let Some(h1) = select_first(document, "h1") {
        let text = text_of_element(&h1);
        if !text.is_empty() {
            return Some(text);
        }
    }

    if let Some(meta) = select_first(document, r#"meta[property="og:title"]"#) {
        if let Some(content) = meta.value().attr("content") {
            if !content.trim().is_empty() {
                return Some(content.trim().to_string());
            }
        }
    }

    select_first(document, "title").map(|title| text_of_element(&title))
}

fn extract_stock_from_buttons(document: &Html) -> (Option<String>, Option<String>) {
    let mut add_to_cart_element = None;
    let mut out_of_stock_element = None;

    for selector in ["button", "input", "a"] {
        for element in select_all(document, selector).into_iter().take(50) {
            let text = if element.value().name() == "input" {
                element
                    .value()
                    .attr("value")
                    .unwrap_or_default()
                    .to_string()
            } else {
                text_of_element(&element)
            };

            if text.is_empty() || text.len() > 80 {
                continue;
            }

            let normalized = text.to_lowercase();
            if ADD_TO_CART_TEXTS
                .iter()
                .any(|pattern| normalized.contains(pattern))
            {
                add_to_cart_element = Some(element);
                break;
            }
            if out_of_stock_element.is_none()
                && OOS_BUTTON_TEXTS
                    .iter()
                    .any(|pattern| normalized.contains(pattern))
            {
                out_of_stock_element = Some(element);
            }
        }

        if add_to_cart_element.is_some() {
            break;
        }
    }

    if let Some(element) = add_to_cart_element {
        if is_element_disabled(&element) {
            return (
                Some("out_of_stock".to_string()),
                Some("button text: disabled".to_string()),
            );
        }
        return (
            Some("in_stock".to_string()),
            Some("button text: enabled".to_string()),
        );
    }

    if out_of_stock_element.is_some() {
        return (
            Some("out_of_stock".to_string()),
            Some("button text: oos".to_string()),
        );
    }

    (None, None)
}

fn result_from_element(element: &ElementRef<'_>) -> Option<ExtractionResult> {
    let text = text_of_element(element);
    if text.is_empty() {
        return None;
    }

    let price = extract_price_from_text(&text);
    let stock_status = extract_stock_from_text(&text);
    if price.is_none() && stock_status.is_none() {
        return None;
    }

    Some(ExtractionResult {
        price,
        stock_status,
        raw: Some(text),
    })
}

fn text_overlaps(stored: &str, candidate: &str) -> bool {
    let stored = normalize_text(stored);
    let candidate = normalize_text(candidate);
    if stored.is_empty() || candidate.is_empty() {
        return false;
    }

    if PRICE_PATTERN.is_match(&stored) {
        return PRICE_PATTERN.is_match(&candidate);
    }

    stored.contains(&candidate) || candidate.contains(&stored)
}

fn ancestor_tags_score(stored: &[String], candidate: &ElementRef<'_>) -> usize {
    let mut ancestors = Vec::new();
    let mut parent = candidate.parent();
    while let Some(node) = parent {
        if let Some(element) = ElementRef::wrap(node) {
            let name = element.value().name();
            if name != "html" {
                ancestors.insert(0, name.to_string());
            }
            if ancestors.len() >= 10 {
                break;
            }
        }
        parent = node.parent();
    }

    stored
        .iter()
        .rev()
        .zip(ancestors.iter().rev())
        .take_while(|(stored, candidate)| stored == candidate)
        .count()
}
